Bottom: a53d84ab51d2b659a1e251f99ebafa1ef604a6b9
Top:    48aa42f6ef11dd6a4790b9025a9d58b751367964
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-29 16:39:11 +0800

Refresh of vhost-introduce-helper-to

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 06e084f..b7dfab0 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2439,11 +2439,14 @@ EXPORT_SYMBOL_GPL(vhost_dequeue_msg);
 
 /* Prefetch descriptor indices */
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
+				struct vring_used_elem *heads,
 				__virtio16 *indices, u16 num)
 {
-	int ret = 0;
-	u16 last_avail_idx, total;
+	int ret, ret2;
+	u16 last_avail_idx, last_used_idx, total, copied;
 	__virtio16 avail_idx;
+	struct vring_used_elem __user *used;
+	int i;
 
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
@@ -2456,16 +2459,44 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	ret = total = min(total, num);
 
 	while (total) {
-		int ret2 = vhost_get_avail(vq, indices[0],
-			   &vq->avail->ring[last_avail_idx & (vq->num - 1)]);
+		copied = vq->num - (last_avail_idx & (vq->num - 1));
+		copied = min(copied, total);
+		ret2 = vhost_copy_from_user(vq, &indices[ret - total],
+			 &vq->avail->ring[last_avail_idx & (vq->num - 1)],
+					copied * sizeof *indices);
 		if (unlikely(ret2)) {
 			vq_err(vq, "Failed to get descriptors\n");
 			return -EFAULT;
 		}
-		vhost_add_used_elem(vq, indices[0], 0, ret - total);
-		--total;
-		++indices;
-		++last_avail_idx;
+
+		last_avail_idx += copied;
+		total -= copied;
+	}
+
+	if (!heads)
+		return ret;
+
+	for (i = 0; i < ret; i++) {
+		heads[i].id = indices[i];
+		heads[i].len = 0;
+	}
+
+	total = ret;
+	last_used_idx = vq->last_used_idx;
+	while (total) {
+		copied = vq->num - (last_used_idx & (vq->num - 1));
+		copied = min(copied, total);
+		ret2 = vhost_copy_to_user(vq,
+				&vq->used->ring[last_used_idx & (vq->num - 1)],
+				&heads[ret - total], copied * sizeof *used);
+
+		if (unlikely(ret2)) {
+			vq_err(vq, "Failed to update used ring!\n");
+			return -EFAULT;
+		}
+
+		total -= copied;
+		last_used_idx += copied;
 	}
 
 	/* Only get avail ring entries after they have been exposed by guest. */
diff --git a/drivers/vhost/vhost.h b/drivers/vhost/vhost.h
index 9c274de..0d8abd1 100644
--- a/drivers/vhost/vhost.h
+++ b/drivers/vhost/vhost.h
@@ -229,6 +229,7 @@ ssize_t vhost_chr_write_iter(struct vhost_dev *dev,
 			     struct iov_iter *from);
 int vhost_init_device_iotlb(struct vhost_dev *d, bool enabled);
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
+                                struct vring_used_elem *heads,
 				__virtio16 *indices, u16 num);
 
 #define vq_err(vq, fmt, ...) do {                                  \
